### PR TITLE
[examples/flax] clip style image-text training example

### PR DESCRIPTION
**Pull request by patil-suraj**
# What does this PR do?

This PR adds an example script to train CLIP style vision-text dual encoder models using pre-trained vision and text encoder.
It supports `CLIP` and `ViT` as image encoder and `BERT` and ROBERTa` (or any other text encoder in flax) as text encoder.

This example uses`torchvision` and torch `Dataloader` for faster pre-processing and data loading. This is important as data loading and processing can become a bottleneck on TPU. This will be revised later to use `datasets`.

### CODE REVIEW

File: examples/research_projects/jax-projects/hybrid_clip/modeling_hybrid_clip.py [hunk 1]
```diff
@@ -0,0 +1,373 @@
...
+
+        if vision_config.model_type == "clip":
+            model.params["vision_model"]["vision_model"] = vision_model.params["vision_model"]
+            model.params["visual_projection"]["kernel"] = vision_model.params["visual_projection"]["kernel"]
+        else:
+            model.params["vision_model"] = vision_model.params
+
+        model.params["text_model"] = text_model.params
+
+        del text_model
```
Comment (patrickvonplaten) [inline examples/research_projects/jax-projects/hybrid_clip/modeling_hybrid_clip.py:370]: Does this deleting really change anything?

File: examples/research_projects/jax-projects/hybrid_clip/run_hybrid_clip.py [hunk 2]
```diff
@@ -0,0 +1,557 @@
...
+    logger.info(f"  Num Epochs = {num_epochs}")
+    logger.info(f"  Instantaneous batch size per device = {training_args.per_device_train_batch_size}")
+    logger.info(f"  Total train batch size (w. parallel & distributed) = {train_batch_size}")
+    logger.info(f"  Total optimization steps = {total_train_steps}")
+
+    train_time = 0
+    # Create sampling rng
+    rng, input_rng = jax.random.split(rng)
+    train_metrics = []
+
```
Comment (patrickvonplaten) [inline examples/research_projects/jax-projects/hybrid_clip/run_hybrid_clip.py:488]: `train_metrics` can be deleted

**Review by patrickvonplaten — APPROVED** [review]
Looks super nice! 

It's hard to be sure that all the modelling code is correct without tests but fine to merge it! Guess we'll see during the sprint if everything works fine and can add tests at a later stage

**Comment by patrickvonplaten** [comment]
And think you might have to run `make style` once

**Comment by patil-suraj** [comment]
Merging!

**Comment by acul3** [comment]
hi @patil-suraj i run your scrip(run_hybrid_clip.py)

after few step i got error

```
it/s/home/acul/env/lib/python3.8/site-packages/transformers/tokenization_utils_base.py:699: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.
  tensor = as_tensor(value)
Epoch ... (1/40):   0%|                                                | 0/40 [02:43<?, ?it/s]
Traceback (most recent call last):                           | 39/462 [02:43<06:48,  1.03it/s]
  File "run_hybrid_clip.py", line 556, in <module>
    main()
  File "run_hybrid_clip.py", line 502, in main
    state, train_metric = p_train_step(state, batch)
  File "/home/acul/env/lib/python3.8/site-packages/jax/_src/traceback_util.py", line 183, in reraise_with_filtered_traceback
    return fun(*args, **kwargs)
  File "/home/acul/env/lib/python3.8/site-packages/jax/_src/api.py", line 1623, in f_pmapped
    for arg in args: _check_arg(arg)
  File "/home/acul/env/lib/python3.8/site-packages/jax/_src/api.py", line 2296, in _check_arg
    raise TypeError(f"Argument '{arg}' of type {type(arg)} is not a valid JAX type.")
jax._src.traceback_util.UnfilteredStackTrace: TypeError: Argument '[[list([1, 1, 1, 1, 
```
i use bert and clip as model vision and text

**Comment by patil-suraj** [comment]
Thank you for reporting! 
Could you please open an issue and post the command that you used?

**Comment by acul3** [comment]
sure 

opening here:
https://github.com/huggingface/transformers/issues/12502

**Pull request merged**